feat: added pagination for promo enrollment requests

lafbelmonte opened the pull request:
-added pagination for promo enrollment requests
-populate in graphql layer

lafbelmonte committed:
feat: added pagination for promo enrollment requests

-added pagination for promo enrollment requests
-populate in graphql layer

ralphcasipe1 commented on koa-crud/src/pagination/index.ts:
```diff
@@ -0,0 +1,41 @@
+    throw new Error(`Invalid first`);
+  }
+
+  const modifiedData = R.map((node: any) => {
+    return R.assoc('rawCursor', node.cursor.buffer, node);
+  })(data);
```
@lafbelmonte  so the `data`'s shape here is this, right?
```typescript
[
  {
     node: {},
     cursor: Buffer
  } 
]
```

In my opinion the data should come from the result of the mongoose model.

lafbelmonte replied:
No sir, that's for adding rawCursor property to the nodes. Reason for this is because the main buffer is nested from the cursor property(mongoose thingy) and it is easier to R.findIndex to just add another property that holds the buffer. Please check my latest commit I removed this function.

ralphcasipe1 commented on koa-crud/src/pagination/index.ts:
```diff
@@ -0,0 +1,41 @@
+import R from 'ramda';
+
+const paginate = (data, first, after) => {
```
Limit the argument to  two only. if it is more than two then make the argument an `object` instead

lafbelmonte replied:
copy sir

ralphcasipe1 commented on koa-crud/src/pagination/index.ts:
```diff
@@ -0,0 +1,41 @@
+
+const paginate = (data, first, after) => {
+  if (first < 0) {
+    throw new Error(`Invalid first`);
+  }
+
+  const modifiedData = R.map((node: any) => {
+    return R.assoc('rawCursor', node.cursor.buffer, node);
+  })(data);
+
+  const startingIndex = R.findIndex(
+    R.propEq('rawCursor', Buffer.from(after, 'base64')),
+  )(modifiedData);
+
+  if (startingIndex < 0) {
+    throw new Error(`Invalid cursor`);
+  }
+
+  const lastIndex = startingIndex + first;
+
+  const nodes = R.slice(startingIndex, lastIndex, modifiedData);
+
+  const edges = R.map((node: any) => {
+    return {
+      node,
+      cursor: node.cursor,
+    };
+  })(nodes);
+
+  return {
+    totalCount: nodes.length,
+    pageInfo: {
+      endCursor: edges[nodes.length - 1].cursor.toString('base64'),
+      hasNextPage: lastIndex < modifiedData.length,
+    },
+    edges,
+  };
+};
+
+export { paginate };
```
Make a unit test for this. Make sure to also cover the if statement in your test

lafbelmonte replied:
copy sir

ralphcasipe1 commented on koa-crud/src/pagination/index.ts:
```diff
@@ -0,0 +1,41 @@
+
+const paginate = (data, first, after) => {
+  if (first < 0) {
+    throw new Error(`Invalid first`);
+  }
+
+  const modifiedData = R.map((node: any) => {
+    return R.assoc('rawCursor', node.cursor.buffer, node);
+  })(data);
+
+  const startingIndex = R.findIndex(
+    R.propEq('rawCursor', Buffer.from(after, 'base64')),
+  )(modifiedData);
+
+  if (startingIndex < 0) {
+    throw new Error(`Invalid cursor`);
+  }
+
+  const lastIndex = startingIndex + first;
+
+  const nodes = R.slice(startingIndex, lastIndex, modifiedData);
+
+  const edges = R.map((node: any) => {
+    return {
+      node,
+      cursor: node.cursor,
+    };
+  })(nodes);
+
+  return {
+    totalCount: nodes.length,
+    pageInfo: {
+      endCursor: edges[nodes.length - 1].cursor.toString('base64'),
+      hasNextPage: lastIndex < modifiedData.length,
+    },
+    edges,
+  };
+};
+
+export { paginate };
```
export it as `default`

lafbelmonte replied:
copy sir

lafbelmonte committed:
feat(lib): implemented paagination for vendors, promos, and members

-implemented paagination for vendors, promos, and members
-added test for paginate function
-fix broken tests for pagination

ralphcasipe1 commented on koa-crud/test/unit/pagination.spec.ts:
```diff
@@ -0,0 +1,100 @@
+
+const chance = new Chance();
+
+describe.only('Pagination Function', () => {
```
Remove `.only`

lafbelmonte replied:
noted sir

ralphcasipe1 commented on koa-crud/src/graphql/type-defs/vendors.ts:
```diff
@@ -2,7 +2,7 @@ import { gql } from 'apollo-server-koa';
 export default gql`
   extend type Query {
-    vendors: [Vendor]
+    vendors(first: Int!, after: String!): VendorConnection
```
`first` and `after` should be optional

lafbelmonte replied:
noted sir

ralphcasipe1 commented on koa-crud/src/graphql/type-defs/promos.ts:
```diff
@@ -29,6 +29,17 @@ export default gql`
+
+  type PromoConnection {
+    totalCount: Int
+    edges: [PromoConnectionEdge]
```
`edges` should be like this `[PromoConnectionEdge]!`

lafbelmonte replied:
noted sir

ralphcasipe1 commented on koa-crud/src/pagination/paginate.ts:
```diff
@@ -0,0 +1,48 @@
+    first: number;
+    after: string;
+  }): Connection<T> {
+    if (first < 0) {
+      throw new Error(`Invalid first`);
+    }
```
Please make a custom error for this.

lafbelmonte replied:
noted sir

ralphcasipe1 commented on koa-crud/src/pagination/paginate.ts:
```diff
@@ -0,0 +1,48 @@
+      R.propEq('cursorBuffer', Buffer.from(after, 'base64')),
+    )(data);
+
+    if (startingIndex < 0) {
+      throw new Error(`Invalid cursor`);
+    }
```
Please make a custom error for invalid cursor

lafbelmonte replied:
noted sir

ralphcasipe1 commented on koa-crud/src/graphql/type-defs/members.ts:
```diff
@@ -12,6 +12,17 @@ export default gql`
+  }
+
+  type MemberConnection {
+    totalCount: Int
```
`totalCount` should be required. if there is no data then return zero (0)

lafbelmonte replied:
noted sir

lafbelmonte committed:
refactor: refactored typescript typings

-refactored typings
-updated imports

lafbelmonte committed:
feature(lib): first and after optional for pagination

lafbelmonte committed:
refactor: removed omit in returning Connection

lafbelmonte committed:
feature(lib): added custom error for pagination inputs

ralphcasipe1 commented on koa-crud/src/data-access/mongoose/members/actions.ts:
```diff
@@ -1,31 +1,63 @@
+} from '../../../lib/mongoose/models/member';
 
-const actions = ({ Member }: { Member: typeof MemberModel }): MembersStore => {
+type Information = {
+  username?: Member['username'];
+  password?: Member['password'];
+  realName?: Member[`realName`];
+  email?: Member[`email`];
+  bankAccount?: Member[`bankAccount`];
+  balance?: Member[`balance`];
+};
```
Can you rename this type to `MemberInformation`, just for the sake of distinction?

And you can refactor this to:
```ts
type MemberInformation = Partial<Omit<Member, 'omit not needed fields'>>
```

lafbelmonte replied:
noted on this

ralphcasipe1 commented on koa-crud/src/data-access/mongoose/vendors/actions.ts:
```diff
@@ -1,38 +1,66 @@
+} from '../../../lib/mongoose/models/vendor';
 
-const actions = ({ Vendor }: { Vendor: typeof VendorModel }): VendorsStore => {
+type Information = {
+  name?: Vendor['name'];
+  type?: Vendor['type'];
+};
```
Can you rename this to `VendorInformation`, to make this distinctive.

lafbelmonte replied:
noted on this

ralphcasipe1 commented on koa-crud/src/entities/member/entity.ts:
```diff
@@ -1,23 +1,43 @@
 import { MissingMemberInformationError } from '../../custom-errors';
 
-const entity = ({ bcrypt }: { bcrypt: typeof bcryptType }) => {
+type Input = {
+  username: Member[`username`];
+  password?: Member[`password`];
+  realName?: Member[`realName`];
+  email?: Member[`email`];
+  bankAccount?: Member[`bankAccount`];
+  balance?: Member[`balance`];
+};
+
+type Output = {
+  username: Member[`username`];
+  password?: Member[`password`];
+  realName?: Member[`realName`];
+  email?: Member[`email`];
+  bankAccount?: Member[`bankAccount`];
+  balance?: Member[`balance`];
+};
```
You can refactor this:
```ts
type Input =  { username: Member['username'] } & Partial<Omit<Member, 'not needed fields'>>
type Output = { username: Member['username'] } & Partial<Omit<Member, 'not needed fields'>>
```

lafbelmonte replied:
noted on this

ralphcasipe1 commented on koa-crud/src/entities/member/entity.ts:
```diff
@@ -1,23 +1,43 @@
     return {
       username,
-      password: password ? await bcrypt.hash(password, 10) : null,
+      password: password ? await bcrypt.hash(password, 10) : undefined,
```
Why did you return this to `undefined`?

lafbelmonte replied:
will replace this to null

ralphcasipe1 commented on koa-crud/src/entities/promo/entity.ts:
```diff
@@ -1,13 +1,43 @@
 } from '../../custom-errors';
 
-const entity = ({ R }: { R: typeof rType }) => {
+type Input = {
+  name: Promo['name'];
+  template: Promo['template'];
+  title: Promo['title'];
+  description: Promo['description'];
+  status: Promo['status'];
+  minimumBalance: Promo['minimumBalance'];
+  requiredMemberFields: Promo['requiredMemberFields'];
+  submitted: Promo['submitted'];
+  enabled: Promo['enabled'];
+};
+
+type Output = {
+  name: Promo['name'];
+  template: Promo['template'];
+  title: Promo['title'];
+  description: Promo['description'];
+  status: Promo['status'];
+  minimumBalance: Promo['minimumBalance'];
+  requiredMemberFields: Promo['requiredMemberFields'];
+  submitted: Promo['submitted'];
+  enabled: Promo['enabled'];
+};
```
You can refactor this to be simpler

lafbelmonte replied:
noted on this

ralphcasipe1 reviewed (changes requested):
Please see my requested changes

lafbelmonte committed:
refactor: simplified typings in input and output

ralphcasipe1 commented on koa-crud/src/data-access/mongoose/members/actions.ts:
```diff
@@ -1,31 +1,58 @@
+  Pick<Member, 'username' | 'realName' | 'email' | 'bankAccount' | 'balance'>
+>;
+
+type MemberFilters = {
+  _id?: Member['_id'] | Record<string, any>;
+  username?: Member['username'] | Record<string, any>;
+};
```
Why do you need this `| Record<string, any>` in both `_id` and `username`

lafbelmonte replied:
I need that for mongoose query operators like $gte, etc.

ralphcasipe1 commented on koa-crud/src/data-access/mongoose/promo-enrollment-requests/actions.ts:
```diff
@@ -1,35 +1,59 @@
+  PromoEnrollmentRequestDocument,
+} from '../../../lib/mongoose/models/promo-enrollment-request';
+
+type PromoEnrollmentRequestFilters = {
+  _id?: PromoEnrollmentRequest['_id'] | Record<string, any>;
+  promo?: PromoEnrollmentRequest['promo'];
+  member?: PromoEnrollmentRequest['member'];
+};
```
You can use Partial and Pick combo here.

lafbelmonte replied:
noted on this sir

ralphcasipe1 commented on koa-crud/src/graphql/resolvers/promo-enrollments.ts:
```diff
@@ -10,91 +8,62 @@ import {
-    };
-  })(nodes);
+const promoEnrollmentRequests = async (
+  obj,
+  args,
```
@lafbelmonte  This should have a type, right?
Especially the `obj` and `args`

lafbelmonte replied:
will add types for that sir

ralphcasipe1 commented on koa-crud/src/use-cases/members/insert-member.ts:
```diff
@@ -1,24 +1,41 @@
+    username: Member[`username`];
+    password: Member[`password`];
+  } & Partial<Pick<Member, 'realName' | 'email' | 'bankAccount' | 'balance'>>;
+  source?;
```
Why there is no type for `source`?

ralphcasipe1 commented on koa-crud/src/use-cases/promos/select-one-promo.ts:
```diff
@@ -1,13 +1,24 @@
+
+type Output = PromoDocument;
+
+export type SelectOnePromoUseCase = (input: Input) => Promise<Output>;
```
Since most of the usecase type is almost the same why not add a generic type for UseCase like this:
```ts
type UseCase<TInput, TOutput> = (input: TInput) => Promise<TOutput>;

export type SelectOnePromoUseCase = UseCase<Input, Output>;
```

lafbelmonte replied:
noted on this

ralphcasipe1 reviewed (approved):
Please see my requested change.
